src/Interpreters/Context.cpp

```diff
@@ -719,28 +749,62 @@ void Context::setUser(const Credentials & credentials, const Poco::Net::SocketAd
 {
     auto lock = getLock();
 
-    client_info.current_user = credentials.getUserName();
-    client_info.current_address = address;
+    try
+    {
+        client_info.current_user = credentials.getUserName();
+        client_info.current_address = address;
+
+    #if defined(ARCADIA_BUILD)
+        /// This is harmful field that is used only in foreign "Arcadia" build.
+        client_info.current_password.clear();
+        if (const auto * basic_credentials = dynamic_cast<const BasicCredentials *>(&credentials))
+            client_info.current_password = basic_credentials->getPassword();
+    #endif
+
+        /// Find a user with such name and check the credentials.
+        auto new_user_id = getAccessControlManager().login(credentials, address.host());
+        auto new_access = getAccessControlManager().getContextAccess(
+            new_user_id, /* current_roles = */ {}, /* use_default_roles = */ true,
+            settings, current_database, client_info);
+
+        user_id = new_user_id;
+        access = std::move(new_access);
+        current_roles.clear();
+        use_default_roles = true;
+
+        setSettings(*access->getDefaultSettings());
+        auto session_log = getSessionLog();
+
+        const auto user = access->getUser();
+        // Check if this is a not an intercluster session, but the real one.
+        if (session_log && user && dynamic_cast<const BasicCredentials *>(&credentials))
+        {
+            auto log_entry = makeSessionLogElement(SESSION_LOGIN_SUCCESS, user, client_info);
+            if (const auto roles_info = access->getRolesInfo())
+                log_entry.roles = roles_info->getEnabledRolesNames();
```

Comment:
I think it's better to move more code away from `Context.cpp`. We could add a function like

```
SessionLog::addLoginEvent(const Context & context)
```

which will collect all the required information from the session context, make a session log element and add it just in one function.

Reply:
Done